### PR TITLE
Hierarchical Caching Refactoring and Fixing TP issue

### PR DESCRIPTION
**Pull request by xiezhq-hermann**
## Motivation
This is a follow-up PR to https://github.com/sgl-project/sglang/pull/2693, primarily focus on two things:
1. Fixing concurrency issue in the case of `tp_size > 1`, which was discussed by https://github.com/sgl-project/sglang/issues/3998 and https://github.com/sgl-project/sglang/pull/4009
2. Introducing layer-wise KV cache loading and synchronization

## Modifications
1. A new completion checking of hierarchical cache activities with order enforcement across TP workers
2. New event and counters to enable layer-wise KV cache loading, more policies to be added
3. Some style changes

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/distributed/parallel_state.py [hunk 1]
```diff
@@ -187,6 +187,7 @@ class GroupCoordinator:
     rank_in_group: int  # rank inside the group
     cpu_group: ProcessGroup  # group for CPU communication
     device_group: ProcessGroup  # group for device communication
+    cache_group: ProcessGroup  # group for cache communication
```
Comment (merrymercy) [inline python/sglang/srt/distributed/parallel_state.py]: can you reuse the cpu group?

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 2]
```diff
@@ -21,11 +23,17 @@ class HiRadixCache(RadixCache):
     def __init__(
         self,
         req_to_token_pool: ReqToTokenPool,
-        token_to_kv_pool: MHATokenToKVPool,
+        token_to_kv_pool: BaseTokenToKVPool,
```
Comment (Edenzzzz) [inline python/sglang/srt/mem_cache/hiradix_cache.py]: token_to_kv_pool_allocator?

File: python/sglang/srt/distributed/parallel_state.py (hunk 1 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/distributed/parallel_state.py]: I think that might work too, the reason I created a new group is to avoid potential interference. If we make sure only one synchronization happen at a time a single group should be good. To be specific, we have two threads running, one scheduler and one compute, there could be a problem if both threads try to synchronize with other processes using the same group. 
Currently in our implementation, it seems compute threads synchronize only through GPU group and the scheduler thread synchronize through CPU group, if we strictly stick to it we should use the same group. It is more of a design choice, what do you think?

File: python/sglang/srt/mem_cache/hiradix_cache.py (hunk 2 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/mem_cache/hiradix_cache.py]: yes I resolved this and will push the update after more testing.

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 3]
```diff
@@ -81,14 +89,20 @@ def inc_hit_count(self, node: TreeNode):
...
-                # clear the reference
-                del self.ongoing_write_through[ack_id]
-            except Exception:
-                break
+        queue_size = torch.tensor(
+            self.cache_controller.ack_write_queue.qsize(), dtype=torch.int
+        )
+        if torch.distributed.get_world_size(group=self.tp_group) > 1:
+            # synchrnoize TP workers to make the same update to radix cache
+            torch.distributed.all_reduce(
```
Comment (Edenzzzz) [inline python/sglang/srt/mem_cache/hiradix_cache.py:97]: Can we use dist.barrier since all queue_sizes should be the same? Thanks.

Comment (xiezhq-hermann) [reply python/sglang/srt/mem_cache/hiradix_cache.py:97]: It is not guaranteed to be the same and that is actually a source of randomness as some workers might finish data transfer faster than the rest.

File: python/sglang/srt/distributed/parallel_state.py (hunk 1 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/distributed/parallel_state.py]: I changed it to reusing the existing CPU group

**Comment by zhyncs** [comment]
Do we have unit test for this?

**Comment by xiezhq-hermann** [comment]
> Do we have unit test for this?

I will add a set of unit tests for hierarchical caching and incorporate them into CI in a separate PR. This PR is a fix for a rather emergent issue of this feature.
